extract async client abstraction.

niklasad1 opened the pull request:
Let's have a round of reviews for this too, it extracts the `shareable` parts of the `ws client` to an `async client abstration`. 
I have called it `core-client` and it's not used by `HTTP` because it can't be extracted to a `sending` and `receiving` end.
That's we have two different transport traits for a `sender` and `receiver` such that they can be used in different tasks.

We can bikeshed on the naming....

Remaining is to expose the client transports from the wrapper crate `jsonrpsee`

I have tested it in [`subxt here`](https://github.com/paritytech/subxt/pull/341)

The benefit that it would be easy to add other transports such as `IPC`, `web-sys` and so on...

niklasad1 committed:
extract ws client to async client

niklasad1 committed:
bring back jsonrpsee-ws-client crate

niklasad1 committed:
new crate core client

niklasad1 committed:
add missing file

niklasad1 committed:
jsonrpsee crate: add core client

niklasad1 committed:
rexport core client

niklasad1 committed:
more re-exports

niklasad1 committed:
downgrade trait bounds

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into extract-async-client

niklasad1 committed:
update version

niklasad1 committed:
fix nits

niklasad1 committed:
send close reason

niklasad1 commented on types/src/traits.rs:
```diff
@@ -127,3 +127,29 @@ tuple_impls! {
+	/// Error.
+	type Error: std::error::Error + Send + Sync;
+
+	/// Send..
```
```suggestion
	/// Send.
```

niklasad1 committed:
Update types/src/traits.rs

maciejhirsz commented on client/core-client/src/lib.rs:
```diff
@@ -1,57 +1,26 @@
+use futures_channel::{mpsc, oneshot};
+use futures_util::{future::Either, SinkExt, StreamExt};
+use jsonrpsee_types::traits::{
+	Client as ClientT, SubscriptionClient as SubscriptionClientT, TransportReceiver, TransportSender,
```
Should we just rename the trait to `ClientT`?

niklasad1 replied:
yeah, sounds good to me.

dvdplm replied:
If we do, we should do it to all of the traits I think.

maciejhirsz reviewed (approved):
Looks good to me, one question.

TarikGul commented on client/ws-client/src/lib.rs:
```diff
@@ -0,0 +1,178 @@
+	///
+	/// You can also prevent the subscription being dropped by calling
+	/// [`Subscription::next()`](crate::types::Subscription) frequently enough such that the buffer capacity doesn't
+	/// exceeds.
```
```suggestion
	/// exceed its limit.
```

Im a little confused by `frequently enough`. Does this mean a user should monitor the subscriptions and as it starts to get close to exceeding its capacity to call `[`Subscription::next()`]`, therefore the implementation on what `frequently` is, is based on the user?

niklasad1 replied:
Maybe it's poorly phrased but it means that if the server produces items much faster than you would call `Subscription::next()` that might end up with the filling buffer and once it exceeds the configured `max_limit`.

In that scenario you need to either call/poll `Subscription::next` more often or increase the buffer max limit, k?

TarikGul replied:
Ahh, okay yea that totally makes sense. 

> more often or increase the buffer max limit

This is what i assumed, just needed some clarification :). Thanks for the explanation

TarikGul reviewed (approved):
lgtm

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into extract-async-client

dvdplm commented on client/core-client/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+license = "MIT"
+repository = "https://github.com/paritytech/jsonrpsee"
+homepage = "https://github.com/paritytech/jsonrpsee"
+documentation = "https://docs.rs/jsonrpsee-ws-client"
```
```suggestion
documentation = "https://docs.rs/jsonrpsee-core-client"
```

dvdplm commented on client/core-client/src/lib.rs:
```diff
@@ -1,57 +1,25 @@
-pub use soketto::handshake::client::Header;
+pub use jsonrpsee_types as types;
 
 /// Wrapper over a [`oneshot::Receiver`](futures::channel::oneshot::Receiver) that reads
```
```suggestion
/// Wrapper over a [`oneshot::Receiver`](futures_channel::oneshot::Receiver) that reads
```

…but `oneshot::Receiver` should be in scope without the link I think?

dvdplm commented on client/ws-client/src/lib.rs:
```diff
@@ -0,0 +1,179 @@
+//!
+//! `jsonrpsee-ws-client` is a [JSON RPC](https://www.jsonrpc.org/specification) WebSocket client library that's is built for `async/await`.
+//!
+//! ## Runtime support
```
```suggestion
//! ## Async runtime support
```

dvdplm commented on client/ws-client/src/lib.rs:
```diff
@@ -0,0 +1,179 @@
+//!
+//! ## Runtime support
+//!
+//! This library uses `tokio` as the runtime and does not support other kinds of runtimes.
```
```suggestion
//! This library uses `tokio` as the runtime and does not support other runtimes.
```

dvdplm commented on client/ws-client/src/lib.rs:
```diff
@@ -0,0 +1,179 @@
+		self
+	}
+
+	/// Set the max number of redirections to perform until a connection is regarded as failed.
```
```suggestion
	/// Set the max number of redirects to follow during the WS handshake until a connection is considered failed.
```

Also: state the default in the docs.

niklasad1 committed:
move 'TEN_MB_CONST' to core

niklasad1 committed:
Update client/core-client/Cargo.toml

Co-authored-by: David <dvdplm@gmail.com>

niklasad1 committed:
Update client/core-client/src/lib.rs

Co-authored-by: David <dvdplm@gmail.com>

niklasad1 committed:
cargo fmt

niklasad1 committed:
Merge remote-tracking branch 'origin/extract-async-client' into extract-async-client

niklasad1 committed:
Update client/ws-client/src/lib.rs

Co-authored-by: David <dvdplm@gmail.com>

niklasad1 committed:
Update client/ws-client/src/lib.rs

Co-authored-by: David <dvdplm@gmail.com>

niklasad1 committed:
move `async-client` core

The `async-client` is hidden behind a new feature flag `async-client`
because it brings in additional dependecies such as tokio rt.

niklasad1 committed:
Merge remote-tracking branch 'origin/extract-async-client' into extract-async-client

niklasad1 committed:
fix docs

niklasad1 committed:
add example how to use "core client"

niklasad1 commented on client/ws-client/src/lib.rs:
````diff
@@ -0,0 +1,169 @@
+///
+/// ```
+#[derive(Clone, Debug)]
+pub struct WsClientBuilder<'a> {
````
this crate could be removed but I kept for it may be easier to use than to use the core client and plug in the actual transport.

niklasad1 commented on core/Cargo.toml:
```diff
@@ -37,6 +37,14 @@ server = [
 	"tokio",
 ]
 client = ["futures-util"]
+async-client = [
```
I had to introduce a feature to avoid bring in `tokio` and some other deps for `http`

niklasad1 committed:
fix build

niklasad1 commented on http-server/Cargo.toml:
```diff
@@ -13,8 +13,8 @@ documentation = "https://docs.rs/jsonrpsee-http-server"
-jsonrpsee-types = { path = "../types", version = "0.6.0" }
-jsonrpsee-core = { path = "../core", version = "0.6.0", features = ["server", "http-helpers"] }
+jsonrpsee-types = { path = "../types", version = "0.6.1" }
+jsonrpsee-core= { path = "../core", version = "0.6.1", features = ["server", "http-helpers"] }
```
```suggestion
jsonrpsee-core = { path = "../core", version = "0.6.1", features = ["server", "http-helpers"] }
```

niklasad1 committed:
Update http-server/Cargo.toml

niklasad1 commented:
@maciejhirsz @dvdplm @TarikGul 

can take you a quick look on this again after some changes?
if no one objects I'll merge it then :)

TarikGul commented on client/transport/Cargo.toml:
```diff
@@ -0,0 +1,41 @@
+    "jsonrpsee-types",
+    "thiserror",
+    "tracing"
+]
```
new line

dvdplm replied:
```suggestion
]

```

dvdplm committed:
Update client/transport/Cargo.toml
